kazoo/recipe/cache.py

```diff
@@ -0,0 +1,382 @@
+"""TreeCache
+
+:Maintainer: Jiangge Zhang <tonyseek@gmail.com>
+:Maintainer: Haochuan Guo <guohaochuan@gmail.com>
+:Maintainer: Tianwen Zhang <mail2tevin@gmail.com>
+:Status: Alpha
+
+A port of the Apache Curator's TreeCache recipe. It builds an in-memory cache
+of a subtree in ZooKeeper and keeps it up-to-date.
+
+See also: http://curator.apache.org/curator-recipes/tree-cache.html
+"""
+
+from __future__ import absolute_import
+
+import os
+import logging
+import contextlib
+import functools
+import operator
+
+from kazoo.exceptions import NoNodeError
+from kazoo.protocol.states import KazooState, EventType
+
+
+logger = logging.getLogger(__name__)
+
+
+class TreeCache(object):
+    """The cache of a ZooKeeper subtree.
+
+    :param client: A :class:`~kazoo.client.KazooClient` instance.
+    :param path: The root path of subtree.
+    """
+
+    STATE_LATENT = 0
+    STATE_STARTED = 1
+    STATE_CLOSED = 2
+
+    def __init__(self, client, path):
+        self._client = client
+        self._root = TreeNode.make_root(self, path)
+        self._state = self.STATE_LATENT
+        self._outstanding_ops = 0
+        self._is_initialized = False
+        self._error_listeners = []
+        self._event_listeners = []
+
+    def start(self):
+        """Starts the cache.
+
+        The cache is not started automatically. You must call this method.
+
+        After a cache started, all changes of subtree will be synchronized
+        from the ZooKeeper server. Events will be fired for those activity.
+
+        See also :meth:`~TreeCache.listen`.
+
+        .. note::
+
+            This method is not thread safety.
+        """
+        if self._state == self.STATE_LATENT:
+            self._state = self.STATE_STARTED
+        else:
+            raise RuntimeError('already started')
```

Comment:
I think its just the base of all kazoo exceptions, so I think using it should be fine (and adding a docstring saying what will happen).